lib/core/service/health_data_manager.dart

```diff
@@ -0,0 +1,544 @@
+
+    for (final point in data) {
+      final DateTime pointStart = DateTime.parse(point.dateFrom);
+      final DateTime pointEnd = DateTime.parse(point.dateTo);
+
+      final DateTime overlapStart = _maxDateTime(pointStart, segmentStart);
+      final DateTime overlapEnd = _minDateTime(pointEnd, segmentEnd);
+
+      if (overlapStart.isBefore(overlapEnd)) {
+        // Calculate the proportion of the data point within this segment
+        final Duration pointDuration = pointEnd.difference(pointStart);
+        final Duration overlapDuration = overlapEnd.difference(overlapStart);
+
+        if (pointDuration.inMilliseconds > 0) {
+          final double proportion =
+              overlapDuration.inMilliseconds / pointDuration.inMilliseconds;
+          final double pointValue =
+              double.tryParse(point.value.toString()) ?? 0.0;
+          totalValue += pointValue * proportion;
+        }
+      }
+    }
+
+    return totalValue;
+  }
+
+  double _aggregateSessionalData(
+    List<AppHealthDataPoint> data,
+    DateTime segmentStart,
+    DateTime segmentEnd,
+  ) {
+    double totalValue = 0.0;
+
+    for (final point in data) {
+      final DateTime pointStart = DateTime.parse(point.dateFrom);
+      final DateTime pointEnd = DateTime.parse(point.dateTo);
+
+      final DateTime overlapStart = _maxDateTime(pointStart, segmentStart);
+      final DateTime overlapEnd = _minDateTime(pointEnd, segmentEnd);
+
+      if (overlapStart.isBefore(overlapEnd)) {
+        final Duration sessionDuration = pointEnd.difference(pointStart);
+        final Duration overlapDuration = overlapEnd.difference(overlapStart);
+
+        // If more than 50% of the session is in this segment, assign full value
+        if (sessionDuration.inMilliseconds > 0 &&
+            overlapDuration.inMilliseconds >
+                sessionDuration.inMilliseconds / 2) {
+          final double pointValue =
+              double.tryParse(point.value.toString()) ?? 0.0;
+          totalValue += pointValue;
+        }
+      }
+    }
+
+    return totalValue;
+  }
+
+  double _aggregateDurationalData(
+    List<AppHealthDataPoint> data,
+    DateTime segmentStart,
+    DateTime segmentEnd,
+  ) {
+    double totalDuration = 0.0;
+
+    for (final point in data) {
+      final DateTime pointStart = DateTime.parse(point.dateFrom);
+      final DateTime pointEnd = DateTime.parse(point.dateTo);
+
+      // For sleep sessions, assign the full duration to the segment
+      // containing the END time
+      // This way "slept 10 hours" appears on the day you wake up,
+      // not distributed
+      if (pointEnd.isAfter(segmentStart) &&
+          (pointEnd.isBefore(segmentEnd) ||
+              pointEnd.isAtSameMomentAs(segmentEnd))) {
+        final Duration sessionDuration = pointEnd.difference(pointStart);
+        totalDuration += sessionDuration.inMinutes.toDouble();
+      }
+    }
+
+    return totalDuration;
+  }
+
+  /// Utility: Get the maximum of two DateTimes
```

Comment:
yo can remove it right. In the future we can move all these extensions to a new file